addons/package_hierarchy/models/stock_quant_package.py

```diff
@@ -0,0 +1,220 @@
+                    values["company_id"] = comparison_recs[0].company_id
+
+            package.location_id = values["location_id"]
+            package.company_id = values.get("company_id")
+            package.owner_id = values["owner_id"]
+
+    @api.depends("parent_id")
+    def _compute_display_name(self):
+        """Compute the display name for a package. Include name of immediate parent."""
+        for package in self:
+            if package.parent_id:
+                package.display_name = "%s/%s" % (package.parent_id.name, package.name)
+            else:
+                package.display_name = package.name
+
+    def _get_contained_quants(self):
+        """Overide to include picks quants of child packages"""
+        return self.env["stock.quant"].search([("package_id", "child_of", self.ids)])
+
+    def _get_move_lines_domain(self):
+        return [
+            "|",
+            ("result_package_id", "child_of", self.ids),
+            ("package_id", "child_of", self.ids),
+        ]
+
+    def get_move_lines(self, aux_domain=None, **kwargs):
+        MoveLines = self.env["stock.move.line"]
+        domain = self._get_move_lines_domain()
+        if aux_domain:
+            domain.extend(aux_domain)
+        return MoveLines.search(domain, **kwargs)
+
+    def action_view_picking(self):
+        """Overide to include picks of child packages"""
+        action = self.env.ref("stock.action_picking_tree_all").read()[0]
+        domain = self._get_move_lines_domain()
+        pickings = self.env["stock.move.line"].search(domain).mapped("picking_id")
+        action["domain"] = [("id", "in", pickings.ids)]
+        return action
+
+    def product_quantities_by_key(self, get_key=lambda q: q.product_id):
+        """This function computes the product quantities the given package grouped by a key
+            Args:
+                get_key: a callable which takes a quant and returns the key
+
+        """
+        res = {}
+        for key, quant_grp in self._get_contained_quants().groupby(get_key):
+            res[key] = sum(quant_grp.mapped("quantity"))
+        return res
+
+    def is_fulfilled_by(self, move_lines):
+        """ Check if a set of packages are fulfilled by a set of move lines"""
+        Precision = self.env["decimal.precision"]
+
+        def get_key(x):
+            return (x.product_id, x.lot_id)
+
+        precision_digits = Precision.precision_get("Product Unit of Measure")
+        pack_qtys = self.product_quantities_by_key(get_key)
+        pack_move_lines = self.get_move_lines(aux_domain=[("id", "in", move_lines.ids)])
+
+        mls_qtys = {}
+        for key, mls_grp in pack_move_lines.groupby(get_key):
+            mls_qtys[key] = sum(mls_grp.mapped("product_qty"))
+
+        # How do we want to deal with the case where the mls have more than the package has?
+        # ATM we say it doesn't fulfill which is not quite right
+        for key in set(pack_qtys.keys()) | set(mls_qtys.keys()):
```

Comment:
this line could be `set(chain(pack_qtys.keys(),mls_qtys.keys()))`

Reply:
Changed as suggested